### PR TITLE
[QOL-9113] ensure that DOCX files are sniffed correctly

### LINKED ISSUE #87: XLSX files are not detected correctly
Hi all, XLSX files are detected as Zip files when _ckanext-s3filestore_ is active. The reason is, that the extension only looks at the first 512 bytes of the uploaded file: https://github.com/qld-gov-au/ckanext-s3filestore/blob/cf0c5bd/ckanext/s3filestore/uploader.py#L545

The part that differentiates a XLSX from a regular Zip comes later (when, depends on the file as well):

```
In [1]: import magic
In [2]: mime = magic.Magic(mime=True)
In [3]: f = open("empty.xlsx", "rb")
In [4]: mime.from_buffer(f.read(512))
Out[4]: 'application/zip'

In [5]: f.seek(0)
Out[5]: 0

In [6]: mime.from_buffer(f.read(1000))
Out[6]: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
```

Since the amount of bytes you have to read before _python-magic_ determines it's a XLSX changes with XLSX size/complexity, you probably have to pass the whole file to be sure it works reliably.

CKAN by default tries to look at the file extension to determine the mimetype (config `ckan.mimetype_guess = file_ext`), or reads the whole file (`ckan.mimetype_guess = file_contents`): https://github.com/ckan/ckan/blob/0a596b8/ckan/lib/uploader.py#L274

What do you think is the best way to fix this? To behave the same way as CKAN does, or stick to `magic.from_buffer` but read the whole file? Performance wise that probably won't do any harm. Or use `magic.from_file/from_descriptor`?

### PR DESCRIPTION
**Pull request by ThrawnCA**
Fixes #87

### CODE REVIEW

File: ckanext/s3filestore/uploader.py [hunk 1]
```diff
@@ -542,8 +542,18 @@ def __init__(self, resource):
             self.mimetype = resource.get('mimetype')
             if not self.mimetype:
                 try:
-                    # 512 bytes should be enough for a mimetype check
-                    self.mimetype = resource['mimetype'] = mime.from_buffer(self.upload_file.read(512))
+                    try:
+                        # Get type from file extension if we recognise it
+                        self.mimetype = mimetypes.guess_type(self.filename, strict=False)[0]
```
Comment (duttonw) [inline ckanext/s3filestore/uploader.py:547]: what happens if someone changes the suffix, should we also not be checking that the file is 'valid' internally as well?

looking at 
https://github.com/ahupp/python-magic/blob/53aa709affef18cb2d1c5930f780dabfc5e4330c/README.md
# recommend using at least the first 2048 bytes, as less can produce incorrect identification

https://docs.python.org/3/library/mimetypes.html#mimetypes.guess_type

One of the good security models we have is that you can't upload an exe under a pdf or csv etc. Should we be removing this?

Comment (ThrawnCA) [reply ckanext/s3filestore/uploader.py:547]: That's in [ckanext-resource-type-validation](https://github.com/qld-gov-au/ckanext-resource-type-validation/), not here.

File: ckanext/s3filestore/tests/test_uploader.py [hunk 2]
```diff
@@ -190,23 +190,24 @@ def _test_dataset(self, private=False, title='Test Dataset', author='test'):
...
-        file_path = os.path.join(os.path.dirname(__file__), 'data.csv')
+        file_path = os.path.join(os.path.dirname(__file__), filename)
         return helpers.call_action(
             'resource_create',
             package_id=dataset['id'],
             upload=FlaskFileStorage(io.open(file_path, 'rb')),
-            url='data.csv')
+            url=filename)
 
     def test_resource_upload(self):
```
Comment (duttonw) [inline ckanext/s3filestore/tests/test_uploader.py:206]: please add a confusion test, where someone tries to upload a pdf that is actually a csv, and a csv as a pdf.

I feel we need to ensure we are file safe for pdf, csv, xls/xlsx as these get opened automatically in the browser (pdf) and xloader/validation plugin tries to read said data and we don't want to allow low probability ddos or other side attacks (or other events on other sites getting files form us) in allowing non valid file's into our storage.

**Review by duttonw — CHANGES_REQUESTED** [review]
couple more tests please.

File: ckanext/s3filestore/uploader.py (hunk 1 above)
Comment (duttonw) [reply ckanext/s3filestore/uploader.py:547]: ah, ok then, this is just setting the correct type. maybe added a comment in the code that this is for type setting, for file type validation, please ensure [ckanext-resource-type-validation](https://github.com/qld-gov-au/ckanext-resource-type-validation/) is added.

File: ckanext/s3filestore/tests/test_uploader.py (hunk 2 above)
Comment (duttonw) [reply ckanext/s3filestore/tests/test_uploader.py:206]: confusion test would still be good, that file type ending is used.
and if ending is missed, it goes through the mime magic and does it best (i.e if for some reason file type is stripped or mangled)

**Review by duttonw — APPROVED** [review]

**Review by chris-randall-qol — APPROVED** [review]

File: ckanext/s3filestore/tests/test_uploader.py (hunk 2 above)
Comment (ThrawnCA) [reply ckanext/s3filestore/tests/test_uploader.py:206]: Those tests exist in `ckanext-resource-type-validation`, which attempts to cover all the bases for mismatched files and types.

Comment (ThrawnCA) [reply ckanext/s3filestore/tests/test_uploader.py:206]: 2048 bytes, see https://github.com/qld-gov-au/ckanext-resource-type-validation/blob/42e28898be66f44b96f4d366725968a9f7195cd2/ckanext/resource_type_validation/resource_type_validation.py#L95

**Pull request merged**